File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.py
class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        m, n = len(heights), len(heights[0])
        minhp, effort = [(0, 0, 0)], [[math.inf]*n for _ in range(m)]
        dirct = [(1,0), (0,1), (-1,0), (0,-1)]
        
        while minhp:
            eff, i, j = heappop(minhp)
            if eff > effort[i][j]:
                continue
            effort[i][j] = eff
            
            for dx,dy in dirct:
                x, y = i+dx, j+dy
                if 0<=x<m and 0<=y<n:
                    eff2 = max(eff, abs(heights[i][j]-heights[x][y]))
                    if eff2<effort[x][y]:
                        effort[x][y] = eff2
                        heappush(minhp, (eff2, x, y))
        
        return effort[-1][-1]

File: 1631-path-with-minimum-effort/1631-path-with-minimum-effort.java
class Solution {
    int[] dx = {1, -1, 0, 0};
    int[] dy = {0, 0, 1, -1};
    
    public int minimumEffortPath(int[][] heights) {
        int n = heights.length;
        int m = heights[0].length;
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a->a[2]));
        pq.add(new int[]{0, 0, 0});
        
        int[][] effort = new int[n][m];
        for(int[] e: effort)
            Arrays.fill(e, Integer.MAX_VALUE);
        
        while(!pq.isEmpty()){
            int i = pq.peek()[0];
            int j = pq.peek()[1];
            int eff = pq.peek()[2];
            pq.poll();
            
            if(eff > effort[i][j])
                continue;
            effort[i][j] = eff; //for heights = [[3]] testcase
            
            for(int d=0; d<4; ++d){
                int x = i+dx[d];
                int y = j+dy[d];
                if(x>=0 && x<n && y>=0 && y<m){
                    int newEff = Math.max(eff, Math.abs(heights[i][j] - heights[x][y]));
                    if(newEff < effort[x][y]){
                        effort[x][y] = newEff;
                        pq.offer(new int[]{x, y, newEff});
                    }
                }
                
            }
        }
        return effort[n-1][m-1];
    }
}